process/sync/storageBootstrap/sovereignChainShardStorageBootstrapper.go

```diff
@@ -19,5 +25,118 @@ func NewSovereignChainShardStorageBootstrapper(shardStorageBootstrapper *shardSt
 	scssb.getScheduledRootHashMethod = scssb.sovereignChainGetScheduledRootHash
 	scssb.setScheduledInfoMethod = scssb.sovereignChainSetScheduledInfo
 
+	scssb.bootstrapper = scssb
 	return scssb, nil
 }
+
+func (ssb *sovereignChainShardStorageBootstrapper) applyCrossNotarizedHeaders(crossNotarizedHeaders []bootstrapStorage.BootstrapHeaderInfo) error {
+	for _, crossNotarizedHeader := range crossNotarizedHeaders {
+		if crossNotarizedHeader.ShardId != core.MainChainShardId {
+			continue
+		}
+
+		extendedHeader, err := process.GetExtendedShardHeaderFromStorage(crossNotarizedHeader.Hash, ssb.marshalizer, ssb.store)
+		if err != nil {
+			return err
+		}
+
+		log.Debug("added cross notarized header in block tracker",
+			"shard", core.MainChainShardId,
+			"round", extendedHeader.GetRound(),
+			"nonce", extendedHeader.GetNonce(),
+			"hash", crossNotarizedHeader.Hash)
+
+		ssb.blockTracker.AddCrossNotarizedHeader(core.MainChainShardId, extendedHeader, crossNotarizedHeader.Hash)
+		ssb.blockTracker.AddTrackedHeader(extendedHeader, crossNotarizedHeader.Hash)
+	}
+
+	return nil
+}
+
+func (ssb *sovereignChainShardStorageBootstrapper) cleanupNotarizedStorage(shardHeaderHash []byte) {
+	log.Debug("sovereign cleanup notarized storage")
+
+	shardHeader, err := process.GetShardHeaderFromStorage(shardHeaderHash, ssb.marshalizer, ssb.store)
+	if err != nil {
+		log.Debug("sovereign shard header is not found in BlockHeaderUnit storage",
+			"hash", shardHeaderHash)
+		return
+	}
+
+	sovereignHeader, castOk := shardHeader.(data.SovereignChainHeaderHandler)
+	if !castOk {
+		log.Warn("sovereignChainShardStorageBootstrapper.cleanupNotarizedStorage",
+			"error", process.ErrWrongTypeAssertion,
+			"expected shard header of type", "SovereignChainHeaderHandler",
+		)
+		return
+	}
+
+	for _, extendedHeaderHash := range sovereignHeader.GetExtendedShardHeaderHashes() {
+		var extendedHeader data.HeaderHandler
```

Comment:
this line can be moved outside of the for.

Reply:
Right, moved it outside